connector/oidc/oidc.go

```diff
@@ -304,9 +315,12 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
 	var email string
 	emailKey := "email"
 	email, found = claims[emailKey].(string)
-	if !found && c.emailKey != "" {
-		emailKey = c.emailKey
+	if (!found || c.overrideClaimMapping) && c.claimMapping.EmailKey != "" {
+		emailKey = c.claimMapping.EmailKey
 		email, found = claims[emailKey].(string)
+		if !found {
```

Comment:
If overriding is enabled and the claim is empty, BUT the original claim is there, this will fail.

Is that the desirable behavior? I can argue for both.

I think we are seriously hitting the limit of declarative configuration with this as both behavior might be desirable, but adding more options would make this even more complicated.

Can you add a test case for this to verify the behavior and detect possible future "regressions"?

Is email the only affected claim?

Reply:
I agree to that. I changed my implementation such that it applies all the properties given in the ClaimMapping if they are present.
The check for `&& c.emailKey != ""` checks that the override only applies for values that are defined.

I think this resolves your input.

Reply:
I think you misunderstood my comment. I'm talking about the _claim_ not the claim mapping. What happens, if there is an email claim with a valid email, but there is no valid claim with the override value (ie. it's missing). In that case, the perfectly valid email is going to overridden by an empty value.

While I can argue that the point of this config is to _always_ override, it can easily lead to problems if the claim turns out to be empty.

So what I suggested is to add a test that verifies that this is the current behavior, so if anything changes we will know. I'm not sure we can do much more with the current config structure. We could add a check to fall back to the original value (if any) if the override claim is empty.

Reply:
Yes, but that's the expected behaviour. If you configure that a claim overrides the `email` claim, then I would expect that there is no fallback. Why should it? This fallback would be very intransparent to the user.

This is the responsibility of the engineer who configures Dex to assure that first, the claim is present within the defined scopes, second that the customised claim contains the necessary information. Isn't it?

**About the test**
Mainly, you would like to catch:
```
"email": "foo.bar@example.com",
```
with configuration:
```
overrideClaimMapping: true,
emailKey:             "custommail",
```
results in `"foo.bar@example.com"`.
But I think that's wrong. It depends on the scope as well.

At the moment, my implementation results in an error if the hasEmailScope is true.
https://github.com/dexidp/dex/blob/e472fe668e81b7d708c990c056e398412a2dba51/connector/oidc/oidc.go#L313
And it default to "", if hasEmailScope is false. That's probably not good because it shouldn't depend on the `email` scope.

This is what I could improve. Mainly, I have to change the check:
```
if !found && ( hasEmailScope || c.emailKey != "" ) {
```
This would assure that in either case, the error is thrown what I prefer because obviously there is going something wrong. But that's an issue that you already have in the code. What if no `email` scope is present and `emailKey` is set but its claim is not present. No error will be thrown and "" is returned.

What's your preferred way? I would go with throwing an error because something doesn't work as expected. Accordingly, I would change the if condition.

Reply:
I'd agree that the more natural expectation here is to **not** fall back to the original claim. However, due to the fact that this "feature" is controlled by a set of different config options, I'm concerned that some day someone would change the implementation leading to random failures.

As I mentioned above, my concern is about _how_ this behavior is configured, not what the actual behavior is (this is an edge case after all). And this is why I suggested adding a test case that make sure that whenever someone attempts to change this behavior, there is a safe guard that stops us. That's what tests are for after all.

IMO adding a new test case that verifies that in case overriding is forced AND the claim is empty AND the original claim is not, we return an empty value is trivial.

Reply:
Hi @sagikazarmark 
I agree to adding a test, I was just not sure what the preferred implementation should be, error or empty string. As you mentioned, I implemented the empty string return for this edge case and added a test `overrideWithMissingCustomEmailClaim` that takes care of that. :)

Thanks for the discussion!